### PR TITLE
Adds experimental support for React Tethys Apps

### PR DESCRIPTION
**Pull request by swainn**
* Adds a minimal API for Tethys Platform
* Adds scaffold for a Tethys App with embedded React frontend

image

### CODE REVIEW

**Comment by swainn** [comment]
@sdc50 and @shawncrawley I'm happy to report that this is ready for review!

**Comment by swainn** [comment]
@sdc50 I fixed the issues with the error page. This is ready for review again.

**Review by gagelarsen — COMMENTED** [review]
Looks good. Although i helped with the development. So I won't `Approve` it.

File: tethys_portal/settings.py [hunk 1]
```diff
@@ -392,3 +395,37 @@
 # Add any additional specified settings to module
 for setting, value in portal_config_settings.items():
     setattr(this_module, setting, value)
+
+# Cookie Settings
```
Comment (sdc50) [inline tethys_portal/settings.py:399]: Why are these commented lines here? Do we need to make a way to add these settings through the the `portal_config.yml` file?

File: tethys_portal/settings.py [hunk 2]
```diff
@@ -392,3 +395,37 @@
...
+# Production only
+# CSRF_COOKIE_SECURE = not DEBUG
+# SESSION_COOKIE_SECURE = not DEBUG
+
+COOKIE_CONFIG = portal_config_settings.pop('COOKIE_CONFIG', {})
+for setting, value in COOKIE_CONFIG.items():
+    setattr(this_module, setting, value)
+
+# CORS Header Settings
+CORS_ALLOWED_ORIGINS = [
```
Comment (sdc50) [inline tethys_portal/settings.py:417]: Does this need to be overridden in production?

**Review by sdc50 — COMMENTED** [review]
What is this file: tethys_cli/scaffold_templates/app_templates/react/package-lock.json_tmpl
It's too long to visualize in GitHub.

**Comment by sdc50** [comment]
> [Coverage Status](https://coveralls.io/builds/50469728)
> 
> Coverage decreased (-0.01%) to 99.989% when pulling **[d3af43e](https://github.com/tethysplatform/tethys/commit/d3af43e05f214dd2c77a2912c401a371f760b039) on tethys-react-apps** into **[8ac1346](https://github.com/tethysplatform/tethys/commit/8ac1346d385ade1900a6809be5ae1c41859dc48c) on master**.

It looks like the `catch_all` in the `register_controller` is missing from tests.

**Comment by sdc50** [comment]
@swainn I'm sorry I took me a long time to review this. It looks like it needs to be rebased now.

File: tethys_apps/base/controller.py [hunk 3]
```diff
@@ -670,4 +675,14 @@ def register_controllers(root_url: str, modules: Union[str, list, tuple], index:
     UrlMap = url_map_maker(root_url)
     url_maps = [UrlMap(**kwargs) for name, kwargs in names.items()]
 
+    # Add a catch all endpoint for any URL following the app's root URL and map it to the index controller
```
[the remark this answers is not in the document]
Comment (swainn) [reply tethys_apps/base/controller.py:678]: No, it maps it to the named controller. I'll updated the comment.

**Comment by swainn** [comment]
> What is this file: tethys_cli/scaffold_templates/app_templates/react/package-lock.json_tmpl
> It's too long to visualize in GitHub.

It is the package_lock.json for the react app scaffold. It is templated, bc the app pacakge name appears a few times in the file and needs to be inserted. Here's a short snippet:

``` json
{
  "name": "{{project}}_frontend",
  "version": "0.0.1",
  "lockfileVersion": 2,
  "requires": true,
  "packages": {
    "": {
      "name": "{{project}}_frontend",
      "version": "1.0.0",
      "license": "ISC",
      "dependencies": {
        "axios": "^0.27.2",
        "bootstrap": "^5.1.3",
        "color": "^4.2.3",
        "css-loader": "^6.5.1",
        "dotenv": "^16.0.1",
        "eslint-config-react-app": "^7.0.1",
        "file-loader": "^6.2.0",
        "ol": "^6.14.1",
        "prismjs": "^1.28.0",
        "react": "^18.0.0",
        "react-bootstrap": "^2.4.0",
        "react-dom": "^18.0.0",
        "react-icons": "^4.3.1",
        "react-plotly.js": "^2.5.1",
        "react-router-bootstrap": "^0.26.1",
        "react-router-dom": "^6.3.0",
        "sass": "^1.49.0",
        "sass-loader": "^12.3.0",
        "style-loader": "^3.3.1",
        "styled-components": "^5.3.5",
        "webpack": "^5.64.4",
        "webpack-cli": "^4.9.1",
        "webpack-dev-server": "^4.6.0"
      },
      "devDependencies": {
        "@babel/core": "^7.18.2",
        "@babel/plugin-proposal-class-properties": "^7.17.12",
        "@babel/plugin-transform-modules-commonjs": "^7.18.2",
        "@babel/preset-env": "^7.18.2",
        "@babel/preset-react": "^7.17.12",
        "@testing-library/jest-dom": "^5.16.4",
        "@testing-library/react": "^13.2.0",
        "@testing-library/user-event": "^14.2.0",
        "babel-jest": "^28.1.0",
        "babel-loader": "^8.2.5",
        "babel-plugin-prismjs": "^2.1.0",
        "babel-plugin-styled-components": "^2.0.7",
        "browserslist": "^4.20.4",
        "dotenv-webpack": "^7.1.0",
        "eslint": "^8.16.0",
        "eslint-plugin-jest": "^26.2.2",
        "eslint-plugin-react": "^7.30.0",
        "eslint-plugin-react-hooks": "^4.5.0",
        "identity-obj-proxy": "^3.0.0",
        "jest": "^28.1.0",
        "jest-css-modules": "^2.1.0",
        "jest-environment-jsdom": "^28.1.0",
        "jest-location-mock": "^1.0.9",
        "jest-watch-typeahead": "^1.1.0",
        "msw": "^0.42.1",
        "react-app-polyfill": "^3.0.0",
        "react-test-renderer": "^18.1.0"
      }
},
    "node_modules/@ampproject/remapping": {
      "version": "2.2.0",
      "resolved": "https://registry.npmjs.org/@ampproject/remapping/-/remapping-2.2.0.tgz",
      "integrity": "sha512-qRmjj8nj9qmLTQXXmaR1cck3UXSRMPrbsLJAasZpF+t3riI71BXed5ebIOYwQntykeZuhjsdweEc9BxH5Jc26w==",
      "dependencies": {
        "@jridgewell/gen-mapping": "^0.1.0",
        "@jridgewell/trace-mapping": "^0.3.9"
      },
      "engines": {
        "node": ">=6.0.0"
      }
    },
    "node_modules/@babel/code-frame": {
      "version": "7.16.7",
      "resolved": "https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.16.7.tgz",
      "integrity": "sha512-iAXqUn8IIeBTNd72xsFlgaXHkMBMt6y4HJp1tIaK465CWLT/fG1aqB7ykr95gHHmlBdGbFeWWfyB4NJJ0nmeIg==",
      "dependencies": {
        "@babel/highlight": "^7.16.7"
      },
      "engines": {
        "node": ">=6.9.0"
      }
    },
    "node_modules/@babel/compat-data": {
      "version": "7.17.10",
      "resolved": "https://registry.npmjs.org/@babel/compat-data/-/compat-data-7.17.10.tgz",
      "integrity": "sha512-GZt/TCsG70Ms19gfZO1tM4CVnXsPgEPBCpJu+Qz3L0LUDsY5nZqFZglIoPC1kIYOtNBZlrnFT+klg12vFGZXrw==",
      "engines": {
        "node": ">=6.9.0"
      }
    },
    "node_modules/@babel/core": {
      "version": "7.18.2",
      "resolved": "https://registry.npmjs.org/@babel/core/-/core-7.18.2.tgz",
      "integrity": "sha512-A8pri1YJiC5UnkdrWcmfZTJTV85b4UXTAfImGmCfYmax4TR9Cw8sDS0MOk++Gp2mE/BefVJ5nwy5yzqNJbP/DQ==",
      "dependencies": {
        "@ampproject/remapping": "^2.1.0",
        "@babel/code-frame": "^7.16.7",
        "@babel/generator": "^7.18.2",
        "@babel/helper-compilation-targets": "^7.18.2",
        "@babel/helper-module-transforms": "^7.18.0",
        "@babel/helpers": "^7.18.2",
        "@babel/parser": "^7.18.0",
        "@babel/template": "^7.16.7",
        "@babel/traverse": "^7.18.2",
        "@babel/types": "^7.18.2",
        "convert-source-map": "^1.7.0",
        "debug": "^4.1.0",
        "gensync": "^1.0.0-beta.2",
        "json5": "^2.2.1",
        "semver": "^6.3.0"
      },
      "engines": {
        "node": ">=6.9.0"
      },
      "funding": {
        "type": "opencollective",
        "url": "https://opencollective.com/babel"
      }
    },
    ...
```

File: tethys_portal/settings.py (hunk 1 above)
Comment (swainn) [reply tethys_portal/settings.py:400]: These are recommended security settings to use for the react setup (from this tutorial https://testdriven.io/blog/django-spa-auth/). However, restricting the CSRF to HTTPONLY means that it cannot be accessed through JavaScript, which a lot of apps take advantage of. This needs more research on our part, so I removed the commented-out lines and added an Issue to look into it: #836

File: tethys_portal/settings.py (hunk 2 above)
Comment (swainn) [reply tethys_portal/settings.py:418]: Probably, though I just tested it without these settings and it seems to work still, so I removed them. We need to create a production configuration tutorial eventually for portals with React apps and recommendations for the CORS settings. See #836 I don't think we need it now for the experimental support.

**Comment by swainn** [comment]
> > [Coverage Status](https://coveralls.io/builds/50469728)
> > Coverage decreased (-0.01%) to 99.989% when pulling **[d3af43e](https://github.com/tethysplatform/tethys/commit/d3af43e05f214dd2c77a2912c401a371f760b039) on tethys-react-apps** into **[8ac1346](https://github.com/tethysplatform/tethys/commit/8ac1346d385ade1900a6809be5ae1c41859dc48c) on master**.
> 
> It looks like the `catch_all` in the `register_controller` is missing from tests.

Good catch. I missed it b/c I have an app installed that has that option defined. I added a test to cover this.

**Review by sdc50 — APPROVED** [review]

**Comment by sdc50** [comment]
@swainn Thanks for all of your work on this! It's looking less likely that I'll get to play with this for my current work project, at least in the near term, but I'm hoping to play around with it at some point.

**Pull request merged**